### PR TITLE
Intly 3612

### PR DESCRIPTION
**Pull request by davidffrench**
## JIRA ID
https://issues.jboss.org/browse/INTLY-3611
https://issues.jboss.org/browse/INTLY-3612
https://issues.jboss.org/browse/INTLY-3613

## Additional Information
Adding the monitoring resource creation to the operator. Since these resource types may not exist on the cluster, this needs to be checked on the cluster before setting up a watch and before reconciling the resources

## Verification Steps
1. Checkout this branch locally
2. Run `make cluster/prepare`
3. Run the operator and watch the logs
4. Run `make cluster/create/examples`
5. Verify the output of the logs does not create any monitoring resources
6. clone this repo - https://github.com/integr8ly/application-monitoring-operator
7. Run `make cluster/install`
8. Verify the operator logs output the watches are setup
9. Verify the 3 monitoring resources have been created.

## Checklist:
- [ ] Verified by team member
- [ ] Comments where necessary
- [ ] Automated Tests
- [ ] Documentation changes if necessary

### CODE REVIEW

**Comment by davidffrench** [comment]
This still needs to have tests added and be more dynamic with the labels and names in the monitoring resources. I am hoping for a review now to get some feedback on it before tomorrow.

**Comment by abstractj** [comment]
@davidffrench please try `go mod tidy` and you should be good to go with GolangCI

File: pkg/controller/keycloak/keycloak_reconciler.go [hunk 1]
```diff
@@ -20,6 +23,18 @@ func NewKeycloakReconciler() *KeycloakReconciler {
 func (i *KeycloakReconciler) Reconcile(clusterState *common.ClusterState, cr *kc.Keycloak) (common.DesiredClusterState, error) {
 	desired := common.DesiredClusterState{}
 	desired = append(desired, i.getKeycloakServiceDesiredState(clusterState, cr))
+
+	// Only add or update the monitoring resources if the resource type exists on the cluster. These booleans are set in the common/autodetect logic
```
[the remark this answers is not in the document]
Comment (davidffrench) [reply pkg/controller/keycloak/keycloak_reconciler.go]: Took a little bit but the viper dependency is now removed

File: pkg/common/cluster_state.go [hunk 2]
```diff
@@ -49,3 +78,78 @@ func (i *ClusterState) readKeycloakServiceCurrentState(context context.Context,
...
+	}
+	return nil
+}
+
+// Keycloak Prometheus Rule. Resource type provided by Prometheus operator
+func (i *ClusterState) readKeycloakPrometheusRuleCurrentState(context context.Context, cr *kc.Keycloak, controllerClient client.Client) error {
+	keycloakPrometheusRule := keycloak.PrometheusRule(cr)
+
+	selector := client.ObjectKey{
+		Name:      keycloakPrometheusRule.Name,
```
[the remark this answers is not in the document]
Comment (davidffrench) [reply pkg/common/cluster_state.go]: Went with the approach from your PR

File: pkg/common/cluster_state.go [hunk 3]
```diff
@@ -15,17 +18,43 @@ import (
...
+	KeycloakServiceMonitor   *monitoringv1.ServiceMonitor
+	KeycloakPrometheusRule   *monitoringv1.PrometheusRule
+	KeycloakGrafanaDashboard *integreatlyv1alpha1.GrafanaDashboard
 }
 
 func NewClusterState() *ClusterState {
 	return &ClusterState{
-		KeycloakService: nil,
+		KeycloakService:          nil,
+		KeycloakServiceMonitor:   nil,
```
[the remark this answers is not in the document]
Comment (davidffrench) [reply pkg/common/cluster_state.go]: Fixed this now

**Comment by davidffrench** [comment]
@malek13 I would love for you to have a look at this as well if you have any free time

**Comment by davidffrench** [comment]
@maleck13 I would love for you to have a look at this as well if you have any free time

**Pull request merged**